agent/consul/autopilot/autopilot.go

```diff
@@ -0,0 +1,457 @@
+		logger:          logger,
+		delegate:        delegate,
+		validServerFunc: serverFunc,
+		interval:        interval,
+		healthInterval:  healthInterval,
+	}
+}
+
+func (a *Autopilot) Start() {
+	a.removeDeadCh = make(chan struct{})
+	a.shutdownCh = make(chan struct{})
+	a.waitGroup = sync.WaitGroup{}
+	a.waitGroup.Add(1)
+
+	go a.run()
+}
+
+func (a *Autopilot) Stop() {
+	close(a.shutdownCh)
+	a.waitGroup.Wait()
+}
+
+// autopilotLoop periodically looks for nonvoting servers to promote and dead servers to remove.
+func (a *Autopilot) run() {
+	defer a.waitGroup.Done()
+
+	// Monitor server health until shutdown
+	ticker := time.NewTicker(a.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-a.shutdownCh:
+			return
+		case <-ticker.C:
+			autopilotConfig, ok := a.delegate.GetOrCreateAutopilotConfig()
+			if !ok {
+				continue
+			}
+
+			// Skip the non-voter promotions unless all servers support the new APIs
+			minRaftProtocol, err := a.MinRaftProtocol()
+			if err != nil {
+				a.logger.Printf("[ERR] autopilot: error getting server raft protocol versions: %s", err)
+				continue
+			}
+			if minRaftProtocol >= 3 {
+				promotions, err := a.delegate.PromoteNonVoters(autopilotConfig, a.GetClusterHealth())
+				if err != nil {
+					a.logger.Printf("[ERR] autopilot: Error checking for non-voters to promote: %s", err)
+				}
+				if err := a.handlePromotions(promotions); err != nil {
+					a.logger.Printf("[ERR] autopilot: Error handling promotions: %s", err)
+				}
+			}
+
+			if err := a.pruneDeadServers(autopilotConfig); err != nil {
+				a.logger.Printf("[ERR] autopilot: Error checking for dead servers to remove: %s", err)
+			}
+		case <-a.removeDeadCh:
+			autopilotConfig, ok := a.delegate.GetOrCreateAutopilotConfig()
+			if !ok {
+				continue
+			}
+
+			if err := a.pruneDeadServers(autopilotConfig); err != nil {
+				a.logger.Printf("[ERR] autopilot: Error checking for dead servers to remove: %s", err)
+			}
+		}
+	}
+}
+
+// fmtServer prints info about a server in a standard way for logging.
+func fmtServer(server raft.Server) string {
+	return fmt.Sprintf("Server (ID: %q Address: %q)", server.ID, server.Address)
+}
+
+// pruneDeadServers removes up to numPeers/2 failed servers
+func (a *Autopilot) pruneDeadServers(conf *Config) error {
+	if !conf.CleanupDeadServers {
+		return nil
+	}
+
+	// Failed servers are known to Serf and marked failed, and stale servers
+	// are known to Raft but not Serf.
+	var failed []string
+	staleRaftServers := make(map[string]raft.Server)
+	raftNode := a.delegate.Raft()
+	future := raftNode.GetConfiguration()
+	if err := future.Error(); err != nil {
+		return err
+	}
+	for _, server := range future.Configuration().Servers {
+		staleRaftServers[string(server.Address)] = server
+	}
+	serfLAN := a.delegate.Serf()
+	for _, member := range serfLAN.Members() {
+		valid, parts := metadata.IsConsulServer(member)
+		if valid {
+			if _, ok := staleRaftServers[parts.Addr.String()]; ok {
```

Comment:
We probably should index by UUID as well. Probably don't want to mix in a behavior change as part of this refactor, but it would be good to capture that in a todo issue for later.